pkg/controller/master-controller-manager/application-definition-synchronizer/controller.go

```diff
@@ -0,0 +1,174 @@
+			return fmt.Errorf("handling deletion of application definition: %w", err)
+		}
+		return nil
+	}
+
+	if err := kuberneteshelper.TryAddFinalizer(ctx, r.masterClient, applicationDef, appkubermaticv1.ApplicationDefinitionSeedCleanupFinalizer); err != nil {
+		return fmt.Errorf("failed to add finalizer: %w", err)
+	}
+
+	applicationDefCreatorGetters := []reconciling.NamedAppKubermaticV1ApplicationDefinitionCreatorGetter{
+		applicationDefCreatorGetter(applicationDef),
+	}
+
+	err := r.syncAllSeeds(log, applicationDef, func(seedClient ctrlruntimeclient.Client, appDef *appkubermaticv1.ApplicationDefinition) error {
+		return reconciling.ReconcileAppKubermaticV1ApplicationDefinitions(ctx, applicationDefCreatorGetters, "", seedClient)
+	})
+	if err != nil {
+		r.recorder.Eventf(applicationDef, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		return fmt.Errorf("reconciled application definition: %s: %w", applicationDef.Name, err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, applicationDef *appkubermaticv1.ApplicationDefinition) error {
+	if kuberneteshelper.HasFinalizer(applicationDef, appkubermaticv1.ApplicationDefinitionSeedCleanupFinalizer) {
+		if err := r.syncAllSeeds(log, applicationDef, func(seedClient ctrlruntimeclient.Client, applicationDef *appkubermaticv1.ApplicationDefinition) error {
+			err := seedClient.Delete(ctx, &appkubermaticv1.ApplicationDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: applicationDef.Name,
+				},
+			})
+
+			return ctrlruntimeclient.IgnoreNotFound(err)
+		}); err != nil {
+			return err
+		}
+
+		if err := kuberneteshelper.TryRemoveFinalizer(ctx, r.masterClient, applicationDef, appkubermaticv1.ApplicationDefinitionSeedCleanupFinalizer); err != nil {
+			return fmt.Errorf("failed to remove application definition finalizer %s: %w", applicationDef.Name, err)
+		}
+	}
+	return nil
+}
+
+func (r *reconciler) syncAllSeeds(log *zap.SugaredLogger, applicationDef *appkubermaticv1.ApplicationDefinition, action func(seedClient ctrlruntimeclient.Client, applicationDef *appkubermaticv1.ApplicationDefinition) error) error {
+	for seedName, seedClient := range r.seedClients {
+		log := log.With("seed", seedName)
+
+		log.Debug("Reconciling application definition with seed")
+
+		err := action(seedClient, applicationDef)
+		if err != nil {
+			return fmt.Errorf("failed syncing application definition %s for seed %s: %w", applicationDef.Name, seedName, err)
+		}
+		log.Debug("Reconciled application definition with seed")
+	}
+	return nil
+}
+
+func applicationDefCreatorGetter(applicationDef *appkubermaticv1.ApplicationDefinition) reconciling.NamedAppKubermaticV1ApplicationDefinitionCreatorGetter {
+	return func() (string, reconciling.AppKubermaticV1ApplicationDefinitionCreator) {
+		return applicationDef.Name, func(a *appkubermaticv1.ApplicationDefinition) (*appkubermaticv1.ApplicationDefinition, error) {
+			a.Name = applicationDef.Name
```

Comment:
The reconciling framework takes care of setting the name [and namespace] already for you :)